cpp/src/arrow/util/int-util.cc

```diff
@@ -402,5 +402,45 @@ void DowncastUInts(const uint64_t* source, uint64_t* dest, int64_t length) {
   memcpy(dest, source, length * sizeof(int64_t));
 }
 
+template <typename InputInt, typename OutputInt>
+void TransposeInts(const InputInt* src, OutputInt* dest, int64_t length,
+                   const int32_t* transpose_map) {
+  while (length >= 4) {
```

Comment:
Clang does this unrolling naturally, gcc does not :( https://godbolt.org/z/ZHobe1

Reply:
Neither does MSVC. The compiler generally doesn't have enough context to know whether unrolling is a good idea (I'm surprised than clang does it at `-O2`, btw). I'd rather unroll explicitly where _we_ have that context.